core/arch/arm/sta/stats.c

```diff
@@ -38,6 +40,80 @@
 			{ 0x87, 0x94, 0x10, 0x02, 0xa5, 0xd5, 0xc6, 0x1b } }
 
 #define STATS_CMD_PAGER_STATS		0
+#define STATS_CMD_ALLOC_STATS		1
+
+#define STATS_NB_POOLS			3
+
+static TEE_Result get_alloc_stats(uint32_t type, TEE_Param p[4])
+{
+	struct tee_mm_pool_stats *stats;
+	uint32_t size_to_retrieve;
+	uint32_t pool_id;
+	uint32_t i;
+
+	/*
+	 * p[0].value.a = pool id (from 0 to n)
+	 *   - 0 means all the pools to be retrieved
+	 *   - 1..n means pool id
+	 * p[0].value.b = 0 if no reset of the stats
+	 * p[1].memref.buffer = output buffer to struct tee_mm_pool_stats
+	 */
+	if (TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+			    TEE_PARAM_TYPE_MEMREF_OUTPUT,
+			    TEE_PARAM_TYPE_NONE,
+			    TEE_PARAM_TYPE_NONE) != type) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	pool_id = p[0].value.a;
+	if (pool_id > STATS_NB_POOLS)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	size_to_retrieve = sizeof(struct tee_mm_pool_stats);
+	if (!pool_id)
+		size_to_retrieve *= STATS_NB_POOLS;
+
+	if (p[1].memref.size < size_to_retrieve) {
+		p[1].memref.size = size_to_retrieve;
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+	p[1].memref.size = size_to_retrieve;
+	stats = (struct tee_mm_pool_stats *)p[1].memref.buffer;
+
+	for (i=1; i<=STATS_NB_POOLS; i++) {
+		if ((pool_id) && (i != pool_id))
+			continue;
+
+		switch (i) {
+		case 1:
+			snprintf(stats->desc, sizeof(stats->desc), "Heap");
+			stats->allocated = malloc_get_allocated();
+			stats->max_allocated = malloc_get_max_allocated();
+			stats->size = malloc_get_heap_size();
+			if (p[0].value.b)
+				malloc_reset_max_allocated();
+			break;
+
+		case 2:
+			snprintf(stats->desc, sizeof(stats->desc), "Public DDR");
+			tee_mm_get_pool_stats(&tee_mm_pub_ddr, stats, !!p[0].value.b);
```

Comment:
Since OP-TEE is multi-threaded `tee_mm_pub_ddr` shouldn't be accessed without holding relevant lock.